File: admin/src/types/relation.ts
export type RelationType = "one_to_one" | "one_to_many" | "many_to_many";
export type Ownership = "source" | "target" | "none";
export type OnDelete = "cascade" | "set_null" | "restrict" | "detach";
export type FetchStrategy = "lazy" | "eager";
export type WriteMode = "diff" | "replace" | "append";

export const RELATION_TYPES: RelationType[] = [
  "one_to_one",
  "one_to_many",
  "many_to_many",
];

export const OWNERSHIP_OPTIONS: Ownership[] = ["source", "target", "none"];
export const ON_DELETE_OPTIONS: OnDelete[] = [
  "cascade",
  "set_null",
  "restrict",
  "detach",
];
export const FETCH_OPTIONS: FetchStrategy[] = ["lazy", "eager"];
export const WRITE_MODE_OPTIONS: WriteMode[] = ["diff", "replace", "append"];

export interface RelationDefinition {
  name: string;
  type: RelationType;
  source: string;
  target: string;
  source_key: string;
  target_key?: string;
  join_table?: string;
  source_join_key?: string;
  target_join_key?: string;
  ownership: Ownership;
  on_delete: OnDelete;
  fetch?: FetchStrategy;
  write_mode?: WriteMode;
}

/** Shape returned by GET /api/_admin/relations */
export interface RelationRow {
  name: string;
  source: string;
  target: string;
  definition: string | RelationDefinition;
  created_at: string;
  updated_at: string;
}

export function parseRelationDefinition(row: RelationRow): RelationDefinition {
  if (typeof row.definition === "string") {
    return JSON.parse(row.definition);
  }
  return row.definition;
}

File: backend/golang/internal/metadata/relation.go
package metadata

type Relation struct {
	Name          string `json:"name"`
	Type          string `json:"type"` // one_to_one, one_to_many, many_to_many
	Source        string `json:"source"`
	Target        string `json:"target"`
	SourceKey     string `json:"source_key"`
	TargetKey     string `json:"target_key,omitempty"`
	JoinTable     string `json:"join_table,omitempty"`
	SourceJoinKey string `json:"source_join_key,omitempty"`
	TargetJoinKey string `json:"target_join_key,omitempty"`
	Ownership     string `json:"ownership"`  // source, target, none
	OnDelete      string `json:"on_delete"`  // cascade, set_null, restrict, detach
	Fetch         string `json:"fetch,omitempty"`      // lazy (default), eager
	WriteMode     string `json:"write_mode,omitempty"` // diff (default), replace, append
}

func (r *Relation) IsManyToMany() bool {
	return r.Type == "many_to_many"
}

func (r *Relation) IsOneToMany() bool {
	return r.Type == "one_to_many"
}

func (r *Relation) IsOneToOne() bool {
	return r.Type == "one_to_one"
}

// DefaultWriteMode returns the write mode, defaulting to "diff".
func (r *Relation) DefaultWriteMode() string {
	if r.WriteMode != "" {
		return r.WriteMode
	}
	return "diff"
}

// DefaultFetch returns the fetch strategy, defaulting to "lazy".
func (r *Relation) DefaultFetch() string {
	if r.Fetch != "" {
		return r.Fetch
	}
	return "lazy"
}
